### PR TITLE
nixos/boot: add option to disable initrd

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
Fix issue #6520 

###### Things done
- [ ] Tested via one or more NixOS tests
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by danbst** [comment]
would it make sense to create the enable option in `stage-1.nix` instead?

**Comment by rnhmjoj** [comment]
@GrahamcOfBorg test boot boot-stage-1

**Comment by rnhmjoj** [comment]
> would it make sense to create the enable option in stage-1.nix instead?

You are right, it makes more sense there.

**Comment by rnhmjoj** [comment]
@GrahamcOfBorg test boot boot-stage-1

File: nixos/modules/system/boot/kernel.nix [hunk 1]
```diff
@@ -187,7 +187,7 @@ in
 
   ###### implementation
 
-  config = mkIf (!config.boot.isContainer) {
+  config = mkIf config.boot.initrd.enable {
```
Comment (danbst) [inline nixos/modules/system/boot/kernel.nix]: is there any particular need to disable this config based on `boot.initrd.enable`? Everything related to initrd will be disabled in that case, why disable rest?

For example, `system.build.kernel`, `boot.kernelParams`, `hardware.firmware` are irrelevant to initrd, right?

cc also @acertain to explain usecase.

Comment (rnhmjoj) [reply nixos/modules/system/boot/kernel.nix]: Yeah, it seems we have to separate the container from initrd logic somehow.

Comment (rnhmjoj) [reply nixos/modules/system/boot/kernel.nix]: I've split the the configuration in two sections and used `mkMerge`. I'm not so sure if this is 100% correct.

**Review by danbst — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Thank you for the review!

**Comment by volth** [comment]
It would be nice to backport to 20.03, so https://github.com/NixOS/mobile-nixos would work with 20.03
cc @samueldr @worldofpeace

**Comment by rnhmjoj** [comment]
I guess it should be safe enough, it's been a month and no reported issues.

**Comment by samueldr** [comment]
> It would be nice to backport to 20.03, so https://github.com/NixOS/mobile-nixos would work with 20.03

While I agree that it *should be safe enough*, doing this on account of Mobile NixOS alone shouldn't be the deciding factor. Right now, to reduce complexity, it is expected that Mobile NixOS is built against the ever-evolving nixos-unstable channel. I do not know if there are other incompatibilities with 20.03, and do not try to keep it backward compatible for now.

Though, with that said, if there are incompatibilities, it would be fine if contributions can be provided, conditional on the imported nixpkgs revision.

I am ±0 with backporting *for* Mobile NixOS, -1 for backporting a change in NixOS past the freeze. Though with that said, this change *is* pretty self-contained and safe. So maybe +1 as far as risk goes (where +1 is not risky at all).

**Comment by wmertens** [comment]
* When I try to do this in nixops, the evaluation fails at https://github.com/NixOS/nixpkgs/blob/fac1d3c4a7321cf068d2677c948f3987086e10e0/nixos/modules/system/activation/top-level.nix#L29. A few lines ahead there's a check for isContainer
* How about adding a little explanation about the option, like `In order to use this, all the modules necessary to mount the root disk need to be compiled into the kernel`

**Pull request merged**